aten/src/ATen/native/sparse/cuda/SparseMatMul.cu

```diff
@@ -0,0 +1,608 @@
+  auto allocator = THCThrustAllocator(globalContext().lazyInitCUDA());
+  auto policy = thrust::cuda::par(allocator).on(stream);
+
+  thrust::for_each(
+      policy,
+      thrust::make_counting_iterator(int64_t(0)),
+      thrust::make_counting_iterator(int64_t(major_dim)),
+      [output_indices_accessor,
+       csr_output_pointers_accessor,
+       major_dim,
+       nnz] __device__(int64_t i) {
+        auto Ap = csr_output_pointers_accessor.data();
+        int64_t* indices_row = output_indices_accessor[0].data();
+
+        for (int jj = Ap[i];  jj < Ap[i + 1]; jj++) {
+          indices_row[jj] = i;
+        }
+      });
+
+  thrust::for_each(
+    policy,
+    thrust::make_counting_iterator(int64_t(0)),
+    thrust::make_counting_iterator(int64_t(csr_output.nnz_)),
+    [output_indices_accessor,
+      csr_output_pointers_accessor,
+      csr_output_ind_accessor,
+      major_dim,
+      nnz] __device__(int64_t i) {
+      int64_t* indices_col = output_indices_accessor[1].data();
+      indices_col[i] = csr_output_ind_accessor[i];
+    });
+}
+
+template <typename scalar_t, bool grad_by_row>
+void sparse_matmul_kernel_grad(Tensor& output, const Tensor& grad, const Tensor& x) {
+  /* 
+    Computes  the backward output  for matrix C = A*B.
+
+    C = A@B 
+      then 
+    A_grad = C_grad @ B^T
+    B_grad = A^T @ C_grad
+
+    if grad_by_row == true:
+      output = x^T @ C_grad 
+    else:
+      output = C_grad @ x^T 
+  */
+  Tensor grad_filled = at::ones(grad.sizes(), grad.options().layout(kStrided));
+  if (grad_by_row) {
+    sparse_sparse_matmul_cuda_kernel<scalar_t>(output, x.transpose(0, 1).coalesce(), grad_filled.to_sparse());
+  } else {
+    sparse_sparse_matmul_cuda_kernel<scalar_t>(output, grad_filled.to_sparse(), x.transpose(0, 1).coalesce());
+  }
+}
+
+} // end anonymous namespace
+
+Tensor sparse_sparse_matmul_cuda(const Tensor& mat1_, const Tensor& mat2_) {
+  TORCH_INTERNAL_ASSERT(mat1_.is_sparse());
+  TORCH_CHECK(mat1_.dim() == 2);
+  TORCH_CHECK(mat2_.dim() == 2);
+
+  TORCH_CHECK(
+      mat1_.size(1) == mat2_.size(0), "mat1 and mat2 shapes cannot be multiplied (",
+      mat1_.size(0), "x", mat1_.size(1), " and ", mat2_.size(0), "x", mat2_.size(1), ")");
+
+  TORCH_CHECK(mat1_.scalar_type() == mat2_.scalar_type(),
+           "mat1 dtype ", mat1_.scalar_type(), " does not match mat2 dtype ", mat2_.scalar_type());
+
+  Tensor output =
+      at::native::empty_sparse({mat1_.size(0), mat2_.size(1)}, mat1_.options());
+
+  AT_DISPATCH_FLOATING_TYPES(mat1_.scalar_type(), "sparse_matmul", [&] {
+    sparse_sparse_matmul_cuda_kernel<scalar_t>(output, mat1_.coalesce(), mat2_.coalesce());
+  });
+  return output;
+}
+
+Tensor sparse_sparse_matmul_backward_cuda(
+    const Tensor& grad,
+    const Tensor& var,
+    int64_t mult_order) {
+  TORCH_CHECK(
+      mult_order == 0 || mult_order == 1,
+      ": mult_order not in [0, 1] at sparse_matmul_backward_cpu function");
+  Tensor output = at::native::empty_like(grad);
```

Comment:
I think the correct thing to have is `empty_like(var)`.
Given C = A * B we are calculating dA = dC * B^T or dB = A^T * dC. dA has same dimensions as A, similarly dB and B, dC and C. dA and dB are `output` here, dC is `grad`, A and B are `var`.